### PR TITLE
[Merged by Bors] - Change `From<Icosphere>` to `TryFrom<Icosphere>`

### LINKED ISSUE #6476: `clippy::fallible-impl-from`  for `From<Icosphere> for Mesh`
## Bevy version

https://github.com/bevyengine/bevy/commit/c019a60b39c5683656025bc9d24a02744aa59dea

## What you did

Ran `cargo +nightly clippy --workspace --all-targets --all-features -Z unstable-options --keep-going -- -Aclippy::type_complexity -Wclippy::nursery -Aclippy::use_self -Aclippy::missing_const_for_fn -Aclippy::redundant_pub_crate -Aclippy::cognitive_complexity -Aclippy::branches_sharing_code -Aclippy::option_if_let_else -Aclippy::useless_let_if_seq -Aclippy::equatable_if_let -Aclippy::iter_with_drain -Aclippy::iter_on_single_items -Aclippy::iter_on_empty_collections -Aclippy::debug_assert_with_mut_call -Aclippy::as_ptr_cast_mut -Aclippy::type_repetition_in_bounds -Aclippy::trait_duplication_in_bounds -Aclippy::future_not_send -Aclippy::imprecise_flops -Aclippy::suboptimal_flops -Wclippy::doc_markdown -Wclippy::redundant_else -Wclippy::match_same_arms -Wclippy::semicolon_if_nothing_returned -Wclippy::explicit_iter_loop -Wclippy::map_flatten -Dwarnings`.

## What went wrong

```
error: consider implementing `TryFrom` instead
   --> crates\bevy_render\src\mesh\shape\icosphere.rs:23:1
    |
23  | / impl From<Icosphere> for Mesh {
24  | |     fn from(sphere: Icosphere) -> Self {
25  | |         if sphere.subdivisions >= 80 {
26  | |             /*
...   |
102 | |     }
103 | | }
    | |_^
    |
    = help: `From` is intended for infallible conversions only. Use `TryFrom` if there's a possibility for the conversion to fail
note: potential failure(s)
   --> crates\bevy_render\src\mesh\shape\icosphere.rs:57:13
    |
57  | / ...   panic!(
58  | | ...       "Cannot create an icosphere of {} subdivisions due to there being too many vertices being generated: {}. (Limited to 65535 vert...
59  | | ...       sphere.subdivisions,
60  | | ...       number_of_resulting_points
61  | | ...   );
    | |_______^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#fallible_impl_from
    = note: `-D clippy::fallible-impl-from` implied by `-D warnings`
    = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### PR DESCRIPTION
**Pull request by 2ne1ugly**
# Objective

- Fixes  #6476

## Solution

- Return error instead of panic through `TryFrom`
- ~~Add `.except()` in examples~~ 
- Add `.unwrap()` in examples

### CODE REVIEW

**Review by TimJentzsch — APPROVED** [review]
Makes sense to me!

File: crates/bevy_render/src/mesh/shape/icosphere.rs [hunk 1]
```diff
@@ -20,8 +20,31 @@ impl Default for Icosphere {
...
+
+impl std::error::Error for FromIcosphereError {}
+
+impl std::fmt::Display for FromIcosphereError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            FromIcosphereError::TooManyVertices {
+                subdivisions,
+                number_of_resulting_points,
+            } => write!(f, "Cannot create an icosphere of {subdivisions} subdivisions due to there being too many vertices being generated: {number_of_resulting_points}. (Limited to 65535 vertices or 79 subdivisions)")
```
Comment (MrGVSV) [inline crates/bevy_render/src/mesh/shape/icosphere.rs:39]: Nit: This crate comes with `thiserror` I believe, so you could also just use that to generate the `Error` and `Display` impls

Comment (2ne1ugly) [reply crates/bevy_render/src/mesh/shape/icosphere.rs:39]: Wasn't quite sure if we were using thiserror extensively
So I decided to make an Error impl instead.
I can convert it to use thiserror instead if it's okay

Comment (MrGVSV) [reply crates/bevy_render/src/mesh/shape/icosphere.rs:39]: I think we should just be consistent. If we don't want to use it, we should probably remove it. If we do want to use it, we should prefer its usage when possible

**Review by alice-i-cecile — APPROVED** [review]

File: examples/ecs/iter_combinations.rs [hunk 2]
```diff
@@ -54,10 +54,13 @@ fn generate_bodies(
...
-    let mesh = meshes.add(Mesh::from(shape::Icosphere {
-        radius: 1.0,
-        subdivisions: 3,
-    }));
+    let mesh = meshes.add(
+        Mesh::try_from(shape::Icosphere {
+            radius: 1.0,
+            subdivisions: 3,
+        })
+        .expect("Error while creating mesh from icosphere"),
```
Comment (alice-i-cecile) [inline examples/ecs/iter_combinations.rs:62]:
```suggestion
        .unwrap(),
```

We have a nice error type; unwrapping will give better user output and is more maintainable.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Great new error type!

The `.expect`s should be replaced by `.unwrap`: the error returned will be better than a simple string and it's easier to maintain and read.

**Review by IceSentry — APPROVED** [review]

**Pull request closed without merging**